cumulus/pallets/xcmp-queue/src/lib.rs

```diff
@@ -525,19 +530,26 @@ impl<T: Config> Pallet<T> {
 			s.last_mut().expect("can't be empty; a new element was just pushed; qed")
 		};
 		let have_active = details.last_index > details.first_index;
-		let appended = have_active &&
-			<OutboundXcmpMessages<T>>::mutate(recipient, details.last_index - 1, |s| {
-				if XcmpMessageFormat::decode_with_depth_limit(MAX_XCM_DECODE_DEPTH, &mut &s[..]) !=
-					Ok(format)
-				{
-					return false
-				}
-				if s.len() + data.len() > max_message_size {
-					return false
-				}
-				s.extend_from_slice(&data[..]);
+		let appended = if have_active {
+			// Check if it exists in the old queue, if not we check the new queue
+			let mut msg = <OutboundXcmpMessages<T>>::take(recipient, details.last_index as u16 - 1)
```

Comment:
Could there be an overlap where it casts a `u64` to a valid `u16` entry that is different?  
Maybe also check that `last_index <= u16::MAX` or rather use `try_into` to convert them.